### ISSUE TITLE
builtin profiling is not accurate enough

### ISSUE THREAD

**Issue by BusyJay** [tikv/tikv#13336]
### What version of TiKV are you using?
Nightly, but should be a problem since 5.x.

### What operating system and CPU are you using?
Linux.

### Steps to reproduce
Use `curl ip:status_port/debug/pprof/profile` to get a flamegraph.

### What did you expect?
A flamegraph with full informations.

### What did happened?
Some frames are missing.

We have met deadlock issue in the past, so https://github.com/tikv/tikv/pull/11205 is introduced to exclude any stacktrace involve glibc, pthread, libgcc from profiling result. Excluding those symbols make the framegraph not accurate enough, we can easily miss key information like syscalls. #13280 is an example why this matters.

Further more, to fix issues on arm, we force enabling frame pointer in release build, which takes more time to compile and has small impact on performance.

It seems current solution has one benefit that we don't depend on other libraries/tools to do CPU profiling, but introduce a lot of overhead and the result is not that perfect and even [dangerous](https://github.com/tikv/tikv/issues/12453).

Hence instead of using or maintaining builtin profiling, I suggest to try external profiling instead. For a basic version, we can make tiup to include a distribution of perftools to do CPU profiling. For contiguous profiling, I suggest to use eBPF instead, which has near unnoticeable performance impact and doesn't require any change to TiKV. There are also successful practices in the community.

**Comment by BusyJay** [comment]
Maybe we should also try https://github.com/grafana/phlare.

**Comment by mornyx** [comment]
> ~Maybe we should also try https://github.com/grafana/phlare.~
> 
> Turns out it relies on application to generate profiling data.

I have an idea but haven't had time to implement it yet: We can choose not to change the existing way TiKV uses pprof-rs, but just provide another profiling implementation inside pprof-rs: that is `perf_event_open` syscall. This avoids additional third-party components. The implementation code can refer to https://github.com/torvalds/linux/blob/master/tools/perf/builtin-record.c (in my opinion, how to deal with multi-threading is the most difficult point when using `perf_event_open`). This also works fine on systems with older kernel versions. On the new version of the kernel, we can automatically use eBPF's perf_event hook to reduce the performance overhead of profiling.

Maybe @YangKeao had a similar idea?

**Comment by YangKeao** [comment]
>  Maybe @YangKeao had a similar idea?

Yes. I even have a draft implementation (frame pointer only, and the ring buffer management still needs some tune). Previously we were using dwarf, and the remote unwinding also needs much work, so I didn't push it forward. Now we are using frame pointer, so I think it's time to just merge the frame pointer implementation :beers: .

I'll try to find the codes, submit the pr and do some observation / benchmark after the busy days :cry: .

For dwarf, there are some exploration to unwind the stack with ebpf in `perf_event_open`: see https://www.polarsignals.com/blog/posts/2022/11/29/profiling-without-frame-pointers/ (actually, I haven't read it :eyes: ). If the compatibility isn't too bad, I also want to try this method. (I personally didn't like using `perf_event_open` with its original "record stack" sampling.)

**Comment by BusyJay** [comment]
Can you explain more how `perf_event_open` is better than current solution? For example, how to solve the first problem (missing samples) described in the issue?

**Comment by YangKeao** [comment]
> Can you explain more how `perf_event_open` is better than current solution? For example, how to solve the first problem (missing samples) described in the issue?

`perf_event_open` is just the syscall used by `perf`. (With frame pointer) the unwind happens in kernel, and don't need things like block list, so no sample will be lost. Using cpu cycles PMU is also more accurate than the software interruption (signal).

Actually, as TiKV is already using frame pointer to get the callchain, maybe we could remove the blocklist :thinking: (the deadlock problems only exist for dwarf unwind) , it could also relief the problem.
